tests/test_c_generator.py

```diff
@@ -96,6 +96,22 @@ def test_complex_decls(self):
         self._assert_ctoc_correct('int test(const char* const* arg);')
         self._assert_ctoc_correct('int test(const char** const arg);')
 
+    def test_atomic_decls(self):
+        self._assert_ctoc_correct('_Atomic(int *) a;')
+        self._assert_ctoc_correct('_Atomic(int) *a;')
+        self._assert_ctoc_correct('_Atomic int *a;')
+        self._assert_ctoc_correct('auto const _Atomic(int *) a;')
```

Comment:
I rebased the PR on top of the changes and provided a fix to typedefs, yet there still is broken stuff apparently. E.g. this line `auto const _Atomic(int *) a;` generates `auto int * _Atomic a;` losing `const`.

Reply:
The issue is that `CGenerator` does not print quals for `Decl`. I.e. the following AST is generated:

```
auto const _Atomic(int *) a;
FileAST(ext=[Decl(name='a',
                  quals=['const'
                        ],
                  storage=['auto'
                          ],
                  funcspec=[
                           ],
                  type=PtrDecl(quals=['_Atomic'
                                     ],
                               type=TypeDecl(declname='a',
                                             quals=[
                                                   ],
                                             type=IdentifierType(names=['int'
                                                                       ]
                                                                 )
                                             )
                               ),
                  init=None,
                  bitsize=None
                  )
            ]
        )
```
and then the generator turns it into:
```
auto int * _Atomic a;

FileAST(ext=[Decl(name='a',
                  quals=[
                        ],
                  storage=['auto'
                          ],
                  funcspec=[
                           ],
                  type=PtrDecl(quals=['_Atomic'
                                     ],
                               type=TypeDecl(declname='a',
                                             quals=[
                                                   ],
                                             type=IdentifierType(names=['int'
                                                                       ]
                                                                 )
                                             )
                               ),
                  init=None,
                  bitsize=None
                  )
            ]
        )
```

An obvious fix is to add `quals` printing in `def _generate_decl(self, n):`.

```diff
diff --git a/pycparser/c_generator.py b/pycparser/c_generator.py
index ded8c65..5a977e5 100644
--- a/pycparser/c_generator.py
+++ b/pycparser/c_generator.py
@@ -418,6 +418,7 @@ class CGenerator(object):
         s = ''
         if n.funcspec: s = ' '.join(n.funcspec) + ' '
         if n.storage: s += ' '.join(n.storage) + ' '
+        if n.quals: s += ' '.join(n.quals) + ' '
         s += self._generate_type(n.type)
         return s
 ```

But that route is a bit complicated as it leads to duplicating quals, e.g. `_Atomic int x;` gives `_Atomic _Atomic int x;` now. This feels wrong in the source, because I see the `_Atomic` qualifier being present twice in the AST:

```
FileAST(ext=[Decl(name='x',
                  quals=['_Atomic'
                        ],
                  storage=[
                          ],
                  funcspec=[
                           ],
                  type=TypeDecl(declname='x',
                                quals=['_Atomic'
                                      ],
                                type=IdentifierType(names=['int'
                                                          ]
                                                    )
                                ),
                  init=None,
                  bitsize=None
                  )
            ]
        )
```

Was the original intention to duplicate quals in both `Decl` and `TypeDecl` but use Decl for user interaction only? This feels wrong, as I cannot put `const` in the first example to e.g. `PtrDecl` `quals`, as it would then refer to the inner type.

Reply:
I faced a very similar issue with `_Alignas(x)` earlier, where I initially put `_Alignas` into quals. To overcome this I added a new `align` attribute here: https://github.com/eliben/pycparser/pull/428/files#diff-1d8b642cdf67bf7baea048a41f5397095407a6cf7768cd9af7356d259d710debR427